src/Block/Fetch.hs

```diff
@@ -57,164 +65,271 @@ import Block.Types (
 
-  let runOgmiosClient = do
-        takeMVar canStart
-        WebSockets.runClient ogmiosAddress ogmiosPort "" $ \wsConn ->
-          runStack $ Right <$> wsApp wsConn blockInfo datumFilter
+-- | Start pair for workers fetching/processing blocks.
+startBlockFetcherAndProcessor ::
+  MonadIO m =>
+  OgmiosInfo ->
+  Hasql.Connection ->
+  BlockInfo ->
+  DatumFilter ->
+  Natural ->
+  m (MVar BlockFetcherEnv, BlockProcessorEnv)
+startBlockFetcherAndProcessor ogmiosInfo dbConn blockInfo datumFilter queueSize = do
+  processorEnv <- mkBlockProcessorEnv datumFilter dbConn queueSize
+  liftIO
+    . void
+    . forkIO
+    . flip runReaderT processorEnv
+    . runStdoutLoggingT
+    . unBlockProcessorApp
+    $ processLoop
+  blockFetcherEnvMVar <- liftIO newEmptyMVar
+  let handleExcpetion (e :: SomeException) = do
+        -- TODO: do we want delay to be configurable?
+        putStr "IO Exception occured, restarting block fetcher in 3s: "
+        print e
+        threadDelay 3_000_000
+        lastBlock' <- runStdoutLoggingT $ getLastBlock dbConn
+        case lastBlock' of
+          Nothing -> runInner blockInfo
+          Just lastBlock -> runInner lastBlock
+      runInner startingBlock = handle handleExcpetion $
+        runClientWith'
+          ogmiosInfo.ogmiosAddress
+          ogmiosInfo.ogmiosPort
+          ""
+          WebSockets.defaultConnectionOptions
+          []
+          $ \wsConn -> do
+            fetcherEnv <- mkBlockFetcherEnv processorEnv wsConn
+            void $ tryPutMVar blockFetcherEnvMVar fetcherEnv
+            void $ swapMVar blockFetcherEnvMVar fetcherEnv
+            flip runReaderT fetcherEnv
+              . runStdoutLoggingT
+              . unBlockFetcherApp
+              $ fetchLoop startingBlock
+  void . liftIO . forkIO . runInner $ blockInfo
+  pure (blockFetcherEnvMVar, processorEnv)
 
-  ogmiosWorker <- liftIO $
-    Async.async $ do
-      runStdoutLoggingT $ logInfoNS "ogmiosWorker" "Starting ogmios client"
-      runOgmiosClient `onException` errorHandler
+------ Fetcher
 
-  putSuccessful <- liftIO $ tryPutMVar envOgmiosWorker $ void ogmiosWorker
-  if putSuccessful
-    then do
-      liftIO $ putMVar canStart ()
-      pure $ Right ()
-    else do
-      Async.cancel ogmiosWorker
-      pure $ Left StartBlockFetcherErrorAlreadyRunning
+newtype BlockFetcherApp a = BlockFetcherApp
+  { unBlockFetcherApp :: LoggingT (ReaderT BlockFetcherEnv IO) a
+  }
+  deriving newtype (Functor, Applicative, Monad, MonadIO, MonadLogger, MonadReader BlockFetcherEnv)
 
-data StopBlockFetcherError
-  = StopBlockFetcherErrorNotRunning
-  deriving stock (Show, Eq)
+-- | Indefinite loop fetching blocks from ogmios and pushing them to the queue.
+fetchLoop ::
+  (MonadIO m, MonadReader BlockFetcherEnv m, MonadLogger m) =>
+  BlockInfo ->
+  m ()
+fetchLoop blockInfo = forever $ do
+  logInfoNS "fetchLoop" "Starting..."
+  void $ findIntersection blockInfo
+  forever fetchNextBlock
```

Comment:
Problem is that it can't be separated because how ogmios itself works, ogmios WS api is stateful so we need to update env every time new intersection is found